### PR TITLE
Raise an exception when the static size of an array field does not match the size of the provided list instead of silently writing incorrectly sized data.

### PR DESCRIPTION
**Pull request by skarndev**
Addresses issue #22.

### CODE REVIEW

File: dissect/cstruct/exceptions.py [hunk 1]
```diff
@@ -12,3 +12,8 @@ class ResolveError(Error):
 
 class NullPointerDereference(Error):
     pass
+
+
+class ArraySizeMismatch(Error):
```
Comment (Schamper) [inline dissect/cstruct/exceptions.py]:
```suggestion
class ArraySizeError(Error):
```
I've been trying to think if there was some Python builtin error we could use but came up blank. I think it's better to rename this to be more consistent with other error names though.

File: dissect/cstruct/types/base.py [hunk 2]
```diff
@@ -3,6 +3,7 @@
 from io import BytesIO
 from typing import Any, BinaryIO, List, TYPE_CHECKING
 
+from dissect.cstruct.exceptions import ArraySizeMismatch
```
Comment (Schamper) [inline dissect/cstruct/types/base.py]:
```suggestion
from dissect.cstruct.exceptions import ArraySizeError
```

File: dissect/cstruct/types/base.py [hunk 3]
```diff
@@ -49,6 +50,9 @@ def dumps(self, data: Any) -> bytes:
 
         Returns:
             The resulting bytes.
+
+        Raises:
+            ArraySizeMismatch: Raised when len(data) does not match the size of a statically sized array field.
```
Comment (Schamper) [inline dissect/cstruct/types/base.py]:
```suggestion
            ArraySizeError: Raised when ``len(data)`` does not match the size of a statically sized array field.
```

File: dissect/cstruct/types/base.py [hunk 4]
```diff
@@ -78,6 +82,9 @@ def write(self, stream: BinaryIO, data: Any) -> int:
 
         Returns:
             The amount of bytes written.
+
+        Raises:
+            ArraySizeMismatch: Raised when len(data) does not match the size of a statically sized array field.
```
Comment (Schamper) [inline dissect/cstruct/types/base.py]:
```suggestion
            ArraySizeError: Raised when ``len(data)`` does not match the size of a statically sized array field.
```

File: dissect/cstruct/types/base.py [hunk 5]
```diff
@@ -158,6 +165,9 @@ def _write(self, stream: BinaryIO, data: List[Any]) -> int:
         if self.null_terminated:
             return self.type._write_0(stream, data)
 
+        if not self.dynamic and self.count != (arr_size := len(data)):
+            raise ArraySizeMismatch(f"Expected static array size {self.count}, got {arr_size} instead.")
```
Comment (Schamper) [inline dissect/cstruct/types/base.py]:
```suggestion
            raise ArraySizeError(f"Expected static array size {self.count}, got {actual_size} instead.")
```

File: dissect/cstruct/types/base.py [hunk 6]
```diff
@@ -158,6 +165,9 @@ def _write(self, stream: BinaryIO, data: List[Any]) -> int:
         if self.null_terminated:
             return self.type._write_0(stream, data)
 
+        if not self.dynamic and self.count != (arr_size := len(data)):
```
Comment (Schamper) [inline dissect/cstruct/types/base.py]:
```suggestion
        if not self.dynamic and self.count != (actual_size := len(data)):
```

File: dissect/cstruct/exceptions.py [hunk 7]
```diff
@@ -12,3 +12,8 @@ class ResolveError(Error):
...
+    pass
+
```
Comment (Schamper) [inline dissect/cstruct/exceptions.py]:
```suggestion
```
Make the linter happy.

File: tests/test_basic.py [hunk 8]
```diff
@@ -417,3 +417,20 @@ def test_dunder_bytes():
...
+    """
+    c = cstruct.cstruct(endian=">")
+    c.load(d)
+
+    a = c.test(a=[1, 2, 3])
+
+    with pytest.raises(ArraySizeMismatch):
+        a.dumps()
+
+
```
Comment (Schamper) [inline tests/test_basic.py]:
```suggestion
```
Make the linter happy.

File: tests/test_basic.py [hunk 9]
```diff
@@ -417,3 +417,20 @@ def test_dunder_bytes():
...
+def test_report_array_size_mismatch():
+    d = """
+    struct test {
+        uint8   a[2];
+    };
+    """
+    c = cstruct.cstruct(endian=">")
+    c.load(d)
+
+    a = c.test(a=[1, 2, 3])
```
Comment (Schamper) [inline tests/test_basic.py]:
```suggestion
    cdef = """
    struct test {
        uint8   a[2];
    };
    """
    cs = cstruct.cstruct(endian=">")
    cs.load(cdef)

    a = cs.test(a=[1, 2, 3])
```

**Review by Schamper — CHANGES_REQUESTED** [review]

**Comment by Schamper** [comment]
@skarnproject thanks, if you could give an ack on the CLA this is good to go!

**Comment by skarndev** [comment]
@DissectBot agree

**Review by Schamper — APPROVED** [review]

**Pull request merged**